### DISCUSSION TITLE
BLE error when compiling with Gitpod

### DISCUSSION THREAD

**Discussion by roesl** [arendst/Tasmota#17162] [Q&A] [answered]
when adding BLE Support in Gitpod following error occurs:

/workspace/Tasmota/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi.ino:63:10: fatal error: NimBLEDevice.h: No such file or directory

**Comment by Jason2866** [comment, marked as answer]
`pio run -e tasmota32-bluetooth`

For MI legacy add this env and run `pio run -e tasmota32-mi32-homebridge`
```
[env:tasmota32-mi32-homebridge]
extends                 = env:tasmota32_base
board_build.f_cpu       = 240000000L
build_flags             = ${env:tasmota32_base.build_flags}
                          -DFIRMWARE_BLUETOOTH
                          -DUSE_MI_EXT_GUI
                          -DUSE_MI_HOMEKIT=1    ; 1 to enable; 0 to disable
                          -DUSE_BERRY_ULP
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_div, lib/lib_ssl
lib_ignore              = ESP8266Audio
                          ESP8266SAM
                          TTGO TWatch Library
                          Micro-RTSP
                          epdiy
```

**Comment by roesl** [comment]
I want also activate PIO and Shutter

**Comment by roesl** [comment]
same error when I run pio run -e tasmota32-bluetooth

**Reply by sfromis** [reply]
Did you try `pio run -e tasmota32-mi32-homebridge` to match the added env?

**Reply by Jason2866** [reply]
Verified `pio run -e tasmota32-bluetooth` and `pio run -e tasmota32-mi32-homebridge` do compile successfully in Gitpod (too).
You are doing something which is breaking compile.

**Reply by roesl** [reply]
pio run -e tasmota32-mi32-homebridge works with BLE
But when I configure PID and Shutter and activate shutter with an 18B20 sensor the slider dont move when I push the up or down buttons and there is no reaction on the temp sensor.
I want to control temperature of a floor heating with a Shelly TRV which I can set to 0-100 % to open valve

**Reply by roesl** [reply]
and every second there is a mqtt command for the shutter values even there is no change of values

**Comment by Jason2866** [comment]
You are not describing exactly what you want to do and how you want to achieve. You are talking in miracles.
The orig. question is fully answered. Imho controlling the Shelly TRV can be easily done in a Smart Home Backend system since mqtt is supported from the TRV.
I see nowhere why BLE should be needed in this use case.

**Reply by roesl** [reply]
I´m want to use PID with Shutter from tasmota and control the Shelly TRV with the value from shutter position via MQTT command.
But shutter and PID of tasmota dont work correct. I want to use a  sample time of about 2-5 minutes to check temperature and write valve position.

**Comment by Jason2866** [comment]
And why BLE? It eats up maaany resources and should never compiled in when
not needed.

roesl ***@***.***> schrieb am Di., 22. Nov. 2022, 14:54:

**Comment by roesl** [comment]
Ble is needed for room temp sensors to control the individual room valves--Diese Nachricht wurde von meinem Android Mobiltelefon mit 1&1 Mail gesendet.Am 22.11.22, 15:39 schrieb Jason2866 ***@***.***>:

**Comment by Jason2866** [comment]
I cant help any further, since you are not explaining the planned setup and how it should work all together.
The use of shutter is still unclear for me. Shelly TRV is not a BLE device afaik.
So for me there are ???  Why not starting to try to controll the device first? If this does work adding other features.

**Comment by roesl** [comment]
Setup is:
   4 MI BLE temp and humidity sensors for rooms. 
   1 MI BLE for outside temp
   ESP32 with tasmota and 18B20 temp sensor for forward temp floor heating with 4 relays for room valves. 
   
   IP-Symcon as Smarthome controller. 
   
   MQTT server runs on IPS. IPS calculates the necessary heating temp depending on outside temp and sets the set temp of PID which compares set and forwardtemp. 
   Valve state is sent to IPS via MQTT and then to Shelly TRV with MQTT or directly from ESP to Shelly TRV with MQTT. 
   
   When possible to run scripts on ESP, the calculation can run there and IPS is not necessary. 
   
   the shutter is for calculating the valve state for the shelly trv
   
   maybe its useful to start a new discussion "floor heating control with tasmota"
The station where the control is for is like this:

![Oventrop Regudis](https://user-images.githubusercontent.com/23452986/203366431-e67a20fe-fff1-4e17-a4a0-415e563ee978.jpg)

  where "Aktor M" = Shelly TRV

**Reply by Scanboostar** [reply]
I ran into the same problem (fatal error: NimBLEDevice.h) after including #define USE_BLE_ESP32 in user_config_override.h to add BLE support for my Tasmota32s3. So I tried the resolution but ran into another problem for both
 pio run -e tasmota32-mi32-homebridge
and
 pio run -e tasmota32-bluetooth

I get: 
/workspace/Tasmota/tasmota/tasmota_xdrv_driver/xdrv_13_display.ino:24:10: fatal error: renderer.h: No such file or directory

I can't find renderer.h

I removed #define USE_BLE_ESP32 from user_config_override.h and the 'normal'compile is successful but
 pio run -e tasmota32-mi32-homebridge
and
 pio run -e tasmota32-bluetooth
still give the same error.

Any Ideas?

**Reply by Scanboostar** [reply]
https://registry.platformio.org/search?q=header:renderer.h
Nothing has been found by your request

Try a less-specific search or use truncation (or wildcard) operator *

**Reply by Jason2866** [reply]
Use a custom env
```
[env:tasmota32s3-mi32-homebridge]
extends                 = env:tasmota32s3
board_build.f_cpu       = 240000000L
build_flags             = ${env:tasmota32_base.build_flags}
                          -DFIRMWARE_BLUETOOTH
                          -DUSE_MI_EXT_GUI
                          -DUSE_MI_HOMEKIT=1    ; 1 to enable; 0 to disable
                          -DUSE_BERRY_ULP
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_div, lib/lib_ssl
lib_ignore              = ESP8266Audio
                          ESP8266SAM
                          TTGO TWatch Library
                          Micro-RTSP
                          epdiy
```

**Reply by Scanboostar** [reply]
I tried but I get the same error:
In file included from tasmota/my_user_config.h:1194,
                 from /workspace/Tasmota/tasmota/tasmota.ino:27:
tasmota/user_config_override.h:5:2: warning: #warning **** user_config_override.h: Using Settings from this File **** [-Wcpp]
 #warning **** user_config_override.h: Using Settings from this File ****
  ^~~~~~~
Retrieved `.pio/build/tasmota32-mi32-homebridge/lib84b/Adafruit_CCS811-1.0.0.14/Adafruit_CCS811.cpp.o' from cache
/workspace/Tasmota/tasmota/tasmota_xdrv_driver/xdrv_13_display.ino:24:10: fatal error: renderer.h: No such file or directory

N.B. I did NOT include #define USE_CCS811

**Reply by Jason2866** [reply]
The env is verified [working](https://github.com/Jason2866/Tasmota-specials/raw/firmware/firmware/tasmota32/other/tasmota32s3-mi32-homebridge.factory.bin). You can not add freely sfuff to the firmware `FIRMWARE_BLUETOOTH`
where the env is based.

**Reply by Scanboostar** [reply]
Without the SSD1306 display config in user_config_override.h it compiles OK, but without a display it is of little use to me. To me it is strange that adding Bluetooth firmware suddenly is causing the requirement of a (non existing) header file related to the display code.

**Comment by Jason2866** [comment]
You have to add the needed lib here `lib_extra_dirs` should be `lib/lib_display`
You have added a display, so the libs for displays are needed. Nothing suddenly.
Changing predefined Tasmota env do need adjusting work. Thats why in Tasmota Docu there is the part.
"Base for all Self Compiled versions has to be Tasmota(32)"
Adding stuff to BLE and WebCam env has a high chance to work not stable. You walk alone.

**Reply by Scanboostar** [reply]
Thanks, It is working now, the compile completed successfully!
(I tried that before but I messed up having duplicate ENV settings in different files ;-<)
